File: GBC.m
function [h,pval]=GBC(R,N,chance,alpha,tail)
	%Implemented by: Shaked Lublinsky and Itay Yaron
	%Preprint: https://osf.io/preprints/psyarxiv/b967v_v1

	%input
	%R: number of correct reponses for each participant
	%N: number of trials for each particiapnt
	%chance: chance prefomance of task
	%alpha: alpha of statistical test 
    %tail: type of alternative hypothesis to evaluate 'both', 'right',
    %'left'

	%output
	%h: 1 test significant, 0 test none significant
	%pval: p-value of the GBC test 


	% checks on inputs
	if size(R,1) >= size(R,2)
		error('R should be a row vector')
	end
	if size(N,1) >= size(N,2)
		error('N should be a row vector')
    end
    if all(size(N) ~= size(R))
        error('N and R should be the same size')
    end
    switch tail
        case 'both'
        case 'right'
        case 'left'
        otherwise 
            error('tail should be assigned a char with the word: both, right or left')
    end

	% calcualte success ratio (SR) from number of correct responses (R) and number of trials (N) per participant
	SR = R ./ N;

	%% Chi square for goodness of fit
	% Correct responses per participant
	A=R';
	% Number of "failures" per participant
	F=N'-R';
	% Expected number of success / failures = number of trials / 2 under the null hypothesis
	E=N'/2;

	% Compute chi-squared statistic: X2 = sum((A - E)^2 / E) + sum((F - E)^2/E),
	% where A indicates number of correct awareness responses, E indicates the
	% expected number of correct responses, and F indicates the number of incorrect responses
	% under the null hyptohesis (50%)
	X2=sum((A-E).^2 ./ E) + sum((F-E).^2 ./ E);

	%Under H0, X2 ~ chisq(k), where degrees of freedom = number of participants
	pvalChi=1 - chi2cdf(X2,length(A));

	%% GB test
	% calculate average standard deviation across all subjects for the sample under the null hyptohesis
	chanceVec=ones(length(A),1)*chance;
	varSample=mean((chanceVec.*(1-chanceVec))./N');
	sigmaSample=sqrt(varSample);
	% z test (note that it is one-sided - 'right' tail)
	[~,pvalGB]=ztest(SR',chance,sigmaSample,'Tail',tail);

	%% GBC test
	%get minimal pvalue between the two tests
	pval=min(pvalChi,pvalGB);
	%we correct for two comparisons so multiply p-value by 2 before
	%comparing with alpha
	pval=pval*2;

	if pval <= alpha
		h=1;
	else
		h=0;
	end
end

File: GBC.R
## GBC test implementation

# Note that JAGS must be installed as a prerequisite.
# Also, note that you must source the AwarenessTests.R script for the function to run
source(file.path('Common', 'AwarenessTests.R'))
if (!require(BSDA)) {
  stop("Package 'rjags' is required but not installed.")
}
if (!require(dplyr)) {
  stop("Package 'rjags' is required but not installed.")
}

# Implemented by: Shaked Lublinsky and Itay Yaron
# Preprint: https://osf.io/preprints/psyarxiv/b967v_v1
GBC <- function(R,N,chance = 0.5,alpha = 0.05, tail = 'right') {
  # Input
  # R: number of correct responses for each participant
  # N: number of trials for each participant
  # chance: chance perfomance of task
  # alpha: alpha of statistical test 
  # tail: type of alternative hypothesis to evaluate 'both', 'right',
  # 'left'
  
  # Output
  # list including:
  # - h: TRUE = significant test result, FALSE = n.s. test result
  # - pval: the p-value of the test 

  #  validate on inputs
  if (!is.null(dim(R))) { stop('R should be a vector') }
  if (!is.null(dim(N))) { stop('N should be a vector') }
  if (length(N) != length(R)) { stop('N and R should be the same length') }
  # validate values
  validate_input_vec <- function(vec, type) {
    if (!is.numeric(vec)) { stop(paste(type, 'Input vector must be a numeric vector')) }
    if (any(is.na(vec))) { stop(paste(type, 'vector contains NA values')) }
    if (any(vec != round(vec))) { stop(paste(type, 'vector must contain only integers.')) }
    
    return(TRUE)
  }
  validate_input_vec(R, 'R')
  validate_input_vec(N, 'N')
  
  # validate tail options
  if (!(tail %in% c('both','right','left'))) { stop('tail must be one of: both, right or left') }
  
  # transform input to gbc_f argument names
  tail <- ifelse(tail == 'right', 'greater', ifelse(tail == 'left', 'less', 'two.sided'))
  AS = R / N
  
  # run test
  pval <- gbc_f(data_as = AS, trials = N, chance = chance, tail = tail)
  # hypothesis testing
  h <- pval <= alpha
  res <- list(pval = pval, h = h)
  
  return(res)
}
